### PR TITLE
Keygen Manager

### PR DESCRIPTION
**Pull request by tbraun96**

### CODE REVIEW

File: dkg-gadget/src/async_protocols/mod.rs [hunk 1]
```diff
@@ -274,12 +274,12 @@ impl OfflinePartyId {
...
-	/// Keygen round is active
-	ACTIVE,
-	/// Keygen round is queued
-	QUEUED,
-	/// UNKNOWN
-	UNKNOWN,
+	/// Executed at the beginning of session 0
+	Genesis,
+	/// Executed at the end of session 0
+	GenesisNext,
```
Comment (drewstone) [inline dkg-gadget/src/async_protocols/mod.rs:280]: Why is this needed? We had Unknown state previously for error cases afaik, CC @shekohex

File: dkg-gadget/src/keygen_manager/mod.rs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+	) {
+		let now_n = header.number().clone();
+		let now: u64 = header.number().saturated_into();
+		let current_protocol = self.session_id_of_active_keygen(now_n);
+		let state = self.state();
+
+		/*
+		   Session 0 (beginning): immediately run genesis
+		   Session 0 (ending): run keygen for session 1
+		   Session 1 (beginning): run keygen for session 2
```
Comment (drewstone) [inline dkg-gadget/src/keygen_manager/mod.rs:112]: This doesn't seem accurate, can you do more investigation, talk with @1xstj. I believe we always are executing the next keygen at the end of the current session. It is tied together with the notion of a session in fact. We only rotate sessions when we successfully generate and sign the next key.

File: dkg-gadget/src/keygen_manager/mod.rs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+		   Session 2 (beginning): run keygen for session 3
+		   Session 3 (beginning): run keygen for session 4
+		*/
+
+		/*
+		   Genesis logic
+		*/
+
+		if now == GENESIS_AUTHORITY_SET_ID && state == KeygenState::Uninitialized {
```
Comment (drewstone) [inline dkg-gadget/src/keygen_manager/mod.rs:121]: Can we split this function into 2? For handling genesis and otherwise? And use `match` statements so its a bit more concise and cleaner?

**Review by drewstone — CHANGES_REQUESTED** [review]

File: dkg-gadget/src/keygen_manager/mod.rs (hunk 3 above)
Comment (tbraun96) [reply dkg-gadget/src/keygen_manager/mod.rs:121]: Sure. The sort of syntactical refactoring you propose usually happens at the very end once the draft goes into a final version. For now, I'm favoring verbosity.

Comment (drewstone) [reply dkg-gadget/src/keygen_manager/mod.rs:131]: Ok

File: dkg-gadget/src/keygen_manager/mod.rs [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+	RunningKeygen,
+	RunningGenesisKeygen,
+	// session_completed denotes the session that executed the keygen, NOT
+	// the generated DKG public key for the next session
+	KeygenCompleted { session_completed: u64 },
+	Failed { session_id: u64 },
+}
+
+// only 1 task at a time may run for keygen
+const MAX_RUNNING_TASKS: usize = 1;
```
Comment (drewstone) [inline dkg-gadget/src/keygen_manager/mod.rs:63]: For these parameters is there any way we can push it out to the CLI/WorkerParams?

File: dkg-gadget/src/keygen_manager/mod.rs [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+use std::{
+	marker::PhantomData,
+	pin::Pin,
+	sync::{
+		atomic::{AtomicUsize, Ordering},
+		Arc,
+	},
+};
+
+pub struct KeygenManager<B: Block, BE, C, GE> {
```
Comment (drewstone) [inline dkg-gadget/src/keygen_manager/mod.rs:28]: This file entirely needs more docs.

File: dkg-gadget/src/keygen_manager/mod.rs [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+			let should_execute_new_keygen = dkg_worker.should_execute_new_keygen(header).await;
+			let executed_count = self.finished_count.load(Ordering::SeqCst);
+			dkg_worker.logger.debug(format!(
+				"*** KeygenManager on_block_finalized: session={session_id},block={block_id}, state={state:?}, current_protocol={current_protocol:?} | total executed: {executed_count}",
+			));
+			dkg_worker
+				.logger
+				.debug(format!("*** Should execute new keygen? {should_execute_new_keygen:?}"));
+
+			/*
```
Comment (drewstone) [inline dkg-gadget/src/keygen_manager/mod.rs:135]: Please use `///` for doc strings like this? Or if you think `//` work better do that. But consistent docs please.

File: dkg-gadget/src/keygen_manager/mod.rs [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+			   Session 1 (ending): run keygen for session 2
+			   Session 2 (ending): run keygen for session 3
+			   Session 3 (ending): run keygen for session 4
+			*/
+
+			/*
+			   Pre checks
+			*/
+
+			// it's possible genesis failed and we need to retry
```
Comment (drewstone) [inline dkg-gadget/src/keygen_manager/mod.rs:147]: Please capitalize doc strings if they're sentences.
```suggestion
			// It's possible genesis failed, and we need to retry
```

File: dkg-gadget/src/keygen_manager/mod.rs [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+			{
+				if dkg_worker.get_dkg_pub_key(header).await.1.is_empty() {
+					dkg_worker.logger.warn(
+						"We will trigger another genesis keygen because the previous one failed",
+					);
+					return self
+						.maybe_start_keygen_for_stage(KeygenRound::Genesis, header, dkg_worker)
+						.await
+				}
+				// if we are at genesis, and genesis keygen is running, do nothing
```
Comment (drewstone) [inline dkg-gadget/src/keygen_manager/mod.rs:159]:
```suggestion
				// If we are at genesis, and genesis keygen is running, do nothing
```

File: dkg-gadget/src/signing_manager/mod.rs [hunk 9]
```diff
@@ -61,149 +72,187 @@ where
...
-						},
-						Err(err) => {
-							dkg_worker
-								.logger
-								.error(format!("Error creating signing protocol: {:?}", &err));
-							dkg_worker.handle_dkg_error(err.clone()).await;
-							return Err(err)
-						},
+			for unsigned_proposal in unsigned_proposals {
+				/*
```
Comment (drewstone) [inline dkg-gadget/src/signing_manager/mod.rs:189]: Wrong format of doc strings.

File: dkg-gadget/src/signing_manager/work_manager.rs [hunk 10]
```diff
@@ -49,43 +70,45 @@ impl<B: BlockT> WorkManager<B> {
...
-
-			let job_receiver = async move {
-				while let Some(task_hash) = rx.recv().await {
-					job_receiver_worker
-						.logger
-						.info_signing(format!("[worker] Received job {task_hash:?}",));
-					job_receiver_worker.poll();
-				}
-			};
+		if let PollMethod::Interval { millis } = poll_method {
```
Comment (drewstone) [inline dkg-gadget/src/signing_manager/work_manager.rs:76]: Docs needed here please.

File: dkg-gadget/src/signing_manager/work_manager.rs [hunk 11]
```diff
@@ -108,12 +131,21 @@ impl<B: BlockT> WorkManager<B> {
...
+			self.to_handler.send(task_hash).map_err(|_| DKGError::GenericError {
+				reason: "Failed to send job to worker".to_string(),
+			})
+		} else {
+			Ok(())
+		}
 	}
 
-	fn poll(&self) {
+	// only relevant for keygen
```
Comment (drewstone) [inline dkg-gadget/src/signing_manager/work_manager.rs:143]: Please add better docs to function signatures.

**Review by drewstone — CHANGES_REQUESTED** [review]
@tbraun96 I'm not seeing enough consistency that I'd expect for you on documentation. Please improve here. The overall impl looks good and the work is great but it's a lot of new abstractions and needs to be properly documented. Also when possible, improve docs of other things being touched here so we get to 100% documented code everywhere.

File: dkg-gadget/src/keygen_manager/mod.rs (hunk 6 above)
Comment (tbraun96) [reply dkg-gadget/src/keygen_manager/mod.rs:135]: `///` for inline code will produce a clippy lint warning. `//` is canonically used for single-line comments. Multi-line comments use the `/* */` pattern. https://doc.rust-lang.org/reference/comments.html

**Comment by drewstone** [comment]
Need your review here @1xstj if we're letting it block your other PRs. Frankly, this doesn't work past the first session so far so more eyes are needed.

**Review by 1xstj — APPROVED** [review]
